src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4237,6 +4237,95 @@ public IAsyncEnumerator<int> GetAsyncEnumerator(System.Threading.CancellationTok
                 );
         }
 
+        [Fact]
+        public void GetAsyncEnumerator_CancellationTokenMustBeOptional_OnIAsyncEnumerable()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+class C
+{
+    public static async Task M(IAsyncEnumerable<int> e)
+    {
+        await foreach (var i in e)
+        {
+        }
+    }
+}
+namespace System.Collections.Generic
+{
+    public interface IAsyncEnumerable<out T>
+    {
+        IAsyncEnumerator<T> GetAsyncEnumerator(System.Threading.CancellationToken token);
+    }
+
+    public interface IAsyncEnumerator<out T> : System.IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask<bool> MoveNextAsync();
+        T Current { get; }
+    }
+}
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(source);
+            comp.VerifyDiagnostics(
+                // (8,33): error CS8411: Asynchronous foreach statement cannot operate on variables of type 'IAsyncEnumerable<int>' because 'IAsyncEnumerable<int>' does not contain a suitable public instance or extension definition for 'GetAsyncEnumerator'
+                //         await foreach (var i in e)
+                Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "e").WithArguments("System.Collections.Generic.IAsyncEnumerable<int>", "GetAsyncEnumerator").WithLocation(8, 33)
+                );
+        }
+
+        [Fact]
+        public void GetAsyncEnumerator_CancellationTokenMustBeOptional_OnIAsyncEnumerable_ExplicitImplementation()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+class C : IAsyncEnumerable<int>
+{
+    public static async Task M(C c)
+    {
+        await foreach (var i in c)
+        {
+        }
+    }
+    IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(System.Threading.CancellationToken token) => throw null;
```

Comment:
>IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(System.Threading.CancellationToken token) => throw null; [](start = 4, length = 119)

It looks like how the interface is implemented makes no difference. Consider adding a test covering implicit implementation as well.

Reply:
It does make a difference, actually. Implicit implementation will error out between lines 757-762 in ForEachLoopBinder.cs, because the implicit implementation should have matched the pattern-based step.

---
In reply to: [550235898](https://github.com/dotnet/roslyn/pull/50105#discussion_r550235898) [](ancestors = 550235898)